### PR TITLE
Deferred Implement

### PR DESCRIPTION
**Pull request by Zangetsu38**
Bassed on other pr, with get smaller cleaner.

### CODE REVIEW

**Comment by Zangetsu38** [comment]
is now ready for review

**Comment by pent0** [comment]
- Commands can now be recorded, for deferred reservations are called to get the buffer that will store these commands
- VDM buffer will be used on all context type when available. For immediate context, when VDM buffer runs out space for new renderer commands, host memory will be allocated.
- VDM buffer allocation order: linear for deferred, pool bitmap allocation for immediate.
- Fix fragment textures and programs not being set back when precomputed is done, fix wrong shader info set for precomputed uniform buffers.

These commits are best to combined with following dynarmic fixes

File: vita3k/modules/SceGxm/SceGxm.cpp [hunk 1]
```diff
@@ -869,37 +1396,58 @@ EXPORT(int, sceGxmEndScene, SceGxmContext *context, const SceGxmNotification *ve
...
+        std::uint8_t *data_allocated = new std::uint8_t[copy_info->source_data_size];
+        std::memcpy(data_allocated, copy_info->source_data, copy_info->source_data_size);
+
+        *copy_info->dest_pointer = data_allocated;
+        copy_info = copy_info->next;
+    }
+
+    // Emit a jump to the first command of given command list
+    // Since only one immediate context exists per process, direct linking like this should be fine! (I hope)
+    renderer::CommandList &imm_cmds = context->renderer->command_list;
```
Comment (sunho) [inline vita3k/modules/SceGxm/SceGxm.cpp:1434]: Don't we need locking here?

Comment (pent0) [reply vita3k/modules/SceGxm/SceGxm.cpp:1434]: Ah that's true

File: vita3k/modules/SceGxm/SceGxm.cpp [hunk 2]
```diff
@@ -825,41 +1299,94 @@ EXPORT(int, sceGxmDrawPrecomputed, SceGxmContext *context, SceGxmPrecomputedDraw
...
+
+    if (!deferredContext->state.active) {
+        return RET_ERROR(SCE_GXM_ERROR_NOT_WITHIN_COMMAND_LIST);
+    }
+
+    // Try to allocate memory for storing command list from renderer
+    int total_to_allocate = (sizeof(renderer::CommandList) + sizeof(renderer::Command) - 1) / sizeof(renderer::Command);
+
+    commandList->copy_info = deferredContext->infos;
+    commandList->list = deferredContext->linearly_allocate<renderer::CommandList>(host.kernel, host.mem, thread_id);
```
Comment (pent0) [inline vita3k/modules/SceGxm/SceGxm.cpp:1344]:
```suggestion
    {
        const std::lock_guard<std::mutex> guard(deferredContext->lock);
        
        commandList->copy_info = deferredContext->infos;
        commandList->list = deferredContext->linearly_allocate<renderer::CommandList>(host.kernel, host.mem, 
            thread_id);
    }
```

File: vita3k/modules/SceGxm/SceGxm.cpp [hunk 3]
```diff
@@ -186,11 +364,161 @@ EXPORT(int, sceGxmAddRazorGpuCaptureBuffer) {
...
+    context->state.viewport.scale.y = -0.5f * static_cast<float>(1.0f + yMax - yMin);
+    context->state.viewport.scale.z = 0.5f;
+
+    context->state.region_clip_min.x = xMin;
+    context->state.region_clip_max.x = xMax;
+    context->state.region_clip_min.y = yMin;
+    context->state.region_clip_max.y = yMax;
+    context->state.region_clip_mode = SCE_GXM_REGION_CLIP_OUTSIDE;
+
+    if (context->alloc_space) {
```
Comment (sunho) [inline vita3k/modules/SceGxm/SceGxm.cpp:384]: When does this thing become null?

Comment (pent0) [reply vita3k/modules/SceGxm/SceGxm.cpp:384]: Just a safe check

Comment (pent0) [reply vita3k/modules/SceGxm/SceGxm.cpp:384]: It may be null sometimes, when deferred context has not start recording. State can still be set before it.

Comment (sunho) [reply vita3k/modules/SceGxm/SceGxm.cpp:384]: ah deferred init it later.

File: vita3k/modules/SceGxm/SceGxm.cpp [hunk 4]
```diff
@@ -639,56 +1010,135 @@ EXPORT(int, sceGxmDisplayQueueFinish) {
...
-        const auto vertex_program = context->record.vertex_program.get(host.mem);
+        const auto vertex_program = context->state.vertex_program.get(host.mem);
         const auto program = vertex_program->program.get(host.mem);
 
         const size_t size = (size_t)program->default_uniform_buffer_count * 4;
         const size_t next_used = context->state.vertex_ring_buffer_used + size;
-        assert(next_used <= context->state.params.vertexRingBufferMemSize);
-        if (next_used > context->state.params.vertexRingBufferMemSize) {
-            return RET_ERROR(SCE_GXM_ERROR_RESERVE_FAILED); // TODO: Does not actually return this on immediate context
+        //assert(next_used <= context->state.vertex_ring_buffer_size);
```
Comment (sunho) [inline vita3k/modules/SceGxm/SceGxm.cpp:1066]: oops

File: vita3k/modules/SceGxm/SceGxm.cpp [hunk 5]
```diff
@@ -414,28 +742,71 @@ EXPORT(int, sceGxmCreateContext, const SceGxmContextParams *params, Ptr<SceGxmCo
     if (!params || !context)
         return RET_ERROR(SCE_GXM_ERROR_INVALID_POINTER);
 
-    *context = alloc<SceGxmContext>(host.mem, __FUNCTION__);
-    if (!*context) {
-        return RET_ERROR(SCE_GXM_ERROR_OUT_OF_MEMORY);
+    if (params->hostMemSize < sizeof(SceGxmContext)) {
```
Comment (sunho) [inline vita3k/modules/SceGxm/SceGxm.cpp:745]: Maybe we should check uniqueness of immeidate context now. Cause, we removed gxm.is_in_scene.

Comment (pent0) [reply vita3k/modules/SceGxm/SceGxm.cpp:745]: Why here?

File: vita3k/renderer/src/gl/draw.cpp [hunk 6]
```diff
@@ -131,8 +131,8 @@ void draw(GLState &renderer, GLContext &context, GxmContextState &state, const F
     glBufferData(GL_ELEMENT_ARRAY_BUFFER, index_size * count, nullptr, GL_DYNAMIC_DRAW);
     glBufferData(GL_ELEMENT_ARRAY_BUFFER, index_size * count, indices, GL_DYNAMIC_DRAW);
 
-    const std::uint8_t *indicies_u8 = reinterpret_cast<const std::uint8_t *>(indices);
-    delete indicies_u8;
+    std::uint8_t *indices_u8 = reinterpret_cast<std::uint8_t *>(indices);
+    delete[] indices_u8;
```
Comment (sunho) [inline vita3k/renderer/src/gl/draw.cpp:135]: This should be moved up to command handler.

Comment (sunho) [reply vita3k/renderer/src/gl/draw.cpp:135]: It's also not protected by deferred context.

File: vita3k/modules/SceGxm/SceGxm.cpp (hunk 5 above)
Comment (sunho) [reply vita3k/modules/SceGxm/SceGxm.cpp:745]: Two contexts might be created, if weird things happen. (e.g. race) It's for debugging purpose. If you don't like the idea, just ignore it.

File: vita3k/renderer/src/gl/draw.cpp (hunk 6 above)
Comment (Zangetsu38) [reply vita3k/renderer/src/gl/draw.cpp:135]: strange, is already like that on first commit 🤔 
image

File: vita3k/mem/include/mem/allocator.h [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+#include <cstdint>
+#include <mutex>
+#include <vector>
+
+struct BitmapAllocator {
+    std::vector<std::uint32_t> words;
+    std::size_t max_offset;
+
+protected:
+    int ForceFill(const std::uint32_t offset, const int size, const bool or_mode = false);
```
Comment (sunho) [inline vita3k/mem/include/mem/allocator.h:12]: Can you make functions snake cases?

Comment (pent0) [reply vita3k/mem/include/mem/allocator.h:12]: I thought most of it is this case

File: vita3k/modules/SceGxm/SceGxm.cpp [hunk 8]
```diff
@@ -25,17 +25,185 @@
...
+    }
+
+    renderer::Command *allocate_new_command(KernelState &kern, const MemState &mem, SceUID current_thread_id) {
+        const std::lock_guard<std::mutex> guard(lock);
+        renderer::Command *new_command = nullptr;
+
+        if (state.type == SCE_GXM_CONTEXT_TYPE_IMMEDIATE) {
+            int size = 1;
+
+            int offset = command_allocator.AllocateFrom(0, size);
```
Comment (sunho) [inline vita3k/modules/SceGxm/SceGxm.cpp:150]: I love this. We're using all memory efficiently.

File: vita3k/renderer/src/gl/draw.cpp (hunk 6 above)
Comment (pent0) [reply vita3k/renderer/src/gl/draw.cpp:135]: I don't understand? All data submitted from command list is only copied, when execute on immediate

Comment (sunho) [reply vita3k/renderer/src/gl/draw.cpp:135]: Oh I see I thought command list can be executed multiple times.

Comment (sunho) [reply vita3k/renderer/src/gl/draw.cpp:135]: Data copy thing was about copying data from pointer at the point of execution, then indexData shouldn't be included there.

Comment (sunho) [reply vita3k/renderer/src/gl/draw.cpp:135]: Never mind about this.

File: vita3k/mem/include/mem/allocator.h (hunk 7 above)
Comment (sunho) [reply vita3k/mem/include/mem/allocator.h:12]: Many of our code is in snake case including your ngs code. Let's try to make them snake case when possible.

**Comment by sunho** [comment]
It generally looks fine for me. I haven't verified bitmap allocator yet, but I guess we can look at it slowly.

**Comment by sunho** [comment]
Tell me when it's ready.

File: vita3k/mem/include/mem/allocator.h (hunk 7 above)
Comment (pent0) [reply vita3k/mem/include/mem/allocator.h:12]: I see. Let's goo

**Comment by pent0** [comment]
> It generally looks fine for me. I haven't verified bitmap allocator yet, but I guess we can look at it slowly.

Feel free lol, I use it in my kernel code for page allocation, so it should be fine already. It's 1to1 copy.

File: vita3k/modules/SceGxm/SceGxm.cpp (hunk 5 above)
Comment (pent0) [reply vita3k/modules/SceGxm/SceGxm.cpp:745]: I'm not sureee, I don't understand why check the uniqueness, since all have same size.

Comment (sunho) [reply vita3k/modules/SceGxm/SceGxm.cpp:745]: I meant checking one immediate context for one processor. Sorry for confusing line reference.

Comment (pent0) [reply vita3k/modules/SceGxm/SceGxm.cpp:745]: This I will do later. Seems to need add some more variables.

**Comment by Zangetsu38** [comment]
> 
> Tell me when it's ready.

All review request is done, normally

**Review by sunho — APPROVED** [review]
Awesome job!

**Comment by Zangetsu38** [comment]
> 
> 
> Awesome job!

sorry, one Command have get deleted by accident, i works on restore it, and is fine after

**Review by pent0 — APPROVED** [review]

**Review by sunho — APPROVED** [review]

**Pull request merged**